src/network/networkd-link.c

```diff
@@ -2588,6 +2587,13 @@ static int link_configure(Link *link) {
                 return 0;
         }
 
+        /* WireGuard always disables link local addressing, since it doesn't fit with
+         * its security model. This is done in the kernel, so it should probably be
+         * reflected here, so that we don't wind up waiting for a link local address
+         * to appear which never appears. */
+        if (streq_ptr(link->kind, "wireguard"))
+                link->network->link_local = ADDRESS_FAMILY_NO;
```

Comment:
The forth line of this function is `assert(link->network)`, so don't be alarmed. This is the right function to do this stuff in, since it's the one that's called just _after_ a network matches and is assigned to `link->network`.

Reply:
Hmm, I am not sure how I feel about this. .network files might be applied to various different interfaces at the same time and some might be wireguard and others might not be. Hence patching around in the network object the first time we apply it to a wireguard iface will have effect on those other interfaces where it probably shouldn't.

I think the much nicer approach would be to patch link_ipv4ll_enabled() in networkd-link.c: it's primarily an accessor to the "link_local" field, but in specific cases will mask ipv4ll support where it's not appropriate. Currently that's on IFF_LOOPBACK devices, but a simple check should be added here to also turns it off for wireguard. (and similar for link_ipv6ll_enabled()).

Reply:
That works for me. I had thought of that initially, but then I saw vcan so I changed directions. I'll force push your suggested change in a few minutes.

Reply:
excellent! thanks!